### PR TITLE
Authentication part2

### LINKED ISSUE #12: Authentication
#### As an application hoster I want to add user authentication to my application.

__Story__:

We have done a great job providing immediate value to our customers. At this
point they can both add and delete images, in addition to a few other features
that are either completed or in the works. Along with a growth in our web
service's popularity have come an increase in number of reports of missing
images. Upon investigation, it turns out that the majority of delete requests
are issued by only a handful of IP addresses. While we could blacklist those IP
addresses from the delete operation, it has always been our plan to only permit
the user who added an image to delete that image. To begin that effort we must
first add the concept of a user.

__Acceptance criteria__:
- [x] Users can sign up.
- [x] Users can log in after they've signed up.
- [x] Users can log out after logging in.
- [x] Users can choose to be remembered as part of the login
  process.<sup>†</sup>
- [x] Multiple browsers can concurrently hold permanent cookies for the same
  user.
- [x] Logged in users should not be able to register, nor log in again.

<sup>†</sup>: "Remembering" means that users will remain logged into the site
even after completely closing their web browser.

__Note__: Do not use Devise for this task.

__Discussion topics__:
- [ ] If we wanted to add more advanced features such as email activation and
  "forgot my password", should we continue to "roll our own" authentication?

__Reference material__:
- https://www.railstutorial.org/book/modeling_users
- https://www.railstutorial.org/book/sign_up
- https://www.railstutorial.org/book/basic_login
- https://www.railstutorial.org/book/advanced_login

__Further reading__:
- https://github.com/plataformatec/devise

__Dependencies__:
- Delete Image
- Refactor flow tests to use page objects

### PR DESCRIPTION
**Pull request by ozhe**
Closes #12

### CODE REVIEW

File: app/helpers/sessions_helper.rb [hunk 1]
```diff
@@ -4,13 +4,25 @@ def log_in(user)
...
   end
 
   def logged_in?
     current_user.present?
   end
 
+  def remember(user)
+    cookies.permanent.signed[:user_id] = user.id
+  end
+
```
Comment (bboe) [inline app/helpers/sessions_helper.rb]: How do you test remember user?

File: app/helpers/sessions_helper.rb [hunk 2]
```diff
@@ -23,7 +23,12 @@ def remember(user)
     cookies.permanent.signed[:user_id] = user.id
   end
 
+  def forget
+    cookies.delete(:user_id)
+  end
+
   def log_out
+    forget
```
Comment (bboe) [inline app/helpers/sessions_helper.rb]: How do you test forget user?

File: app/helpers/sessions_helper.rb [hunk 3]
```diff
@@ -23,7 +23,12 @@ def remember(user)
     cookies.permanent.signed[:user_id] = user.id
   end
 
+  def forget
```
Comment (bboe) [inline app/helpers/sessions_helper.rb]: Is this commit end-to-end if forget is never called?

File: app/helpers/sessions_helper.rb [hunk 4]
```diff
@@ -4,13 +4,25 @@ def log_in(user)
...
+        @current_user = user
+      end
+    end
   end
 
   def logged_in?
     current_user.present?
   end
 
+  def remember(user)
```
Comment (bboe) [inline app/helpers/sessions_helper.rb]: Is this comment end-to-end if remember is not called?

File: test/controllers/sessions_controller_test.rb [hunk 5]
```diff
@@ -52,4 +52,20 @@ class SessionsControllerTest < ActionDispatch::IntegrationTest
...
+
+  test 'log in with remember_me' do
+    user = User.create!(user_params)
+
+    post sessions_path(session_form: session_form_params(remember_me: '1'))
+    assert_equal user.id, session[:user_id]
+    assert_not_nil cookies['user_id']
+  end
+
+  test 'log in without remember_me' do
```
Comment (bboe) [inline test/controllers/sessions_controller_test.rb]: Is this test redundant? Could you put the crucial assertion elsewhere?

File: test/page_objects/sessions/new_page.rb [hunk 6]
```diff
@@ -0,0 +1,19 @@
+module PageObjects
```
Comment (bboe) [inline test/page_objects/sessions/new_page.rb]: Does this commit provide end-to-end value?

File: test/flows/users_log_in_test.rb [hunk 7]
```diff
@@ -0,0 +1,32 @@
+require 'flow_test_helper'
+
+class UsersLogInTest < FlowTestCase
+  teardown do
+    Capybara.reset_sessions!
+  end
+
+  test 'log in an user' do
+    user = User.create!(
```
Comment (bboe) [inline test/flows/users_log_in_test.rb]: Do you have a helper for this process?

File: test/flows/users_sign_up_test.rb [hunk 8]
```diff
@@ -0,0 +1,24 @@
+require 'flow_test_helper'
+
+class UsersSignUpTest < FlowTestCase
+  teardown do
+    Capybara.reset_sessions!
+  end
+
+  test 'Sign up an user' do
```
Comment (bboe) [inline test/flows/users_sign_up_test.rb]: "a user"

File: test/controllers/sessions_controller_test.rb [hunk 9]
```diff
@@ -45,12 +45,31 @@ class SessionsControllerTest < ActionDispatch::IntegrationTest
...
+    post sessions_path(session_form: session_form_params)
+
+    get new_session_path
+    assert_redirected_to images_path
+    assert_equal 'Already logged in', flash[:warning]
+  end
+
+  test 'log in twice' do
+    User.create!(user_params)
+    post sessions_path(session_form: session_form_params)
```
Comment (bboe) [inline test/controllers/sessions_controller_test.rb]: This pattern (create, post) exists in many of these tests. Could we dry that up? It also might make sense to include an assertion that it succeeded just in case it doesn't so you don't get a misleading exception if it were to fail.

File: app/controllers/sessions_controller.rb [hunk 10]
```diff
@@ -13,8 +15,8 @@ def new
   end
 
   def destroy
-    log_out
-    redirect_to images_path
+    log_out if logged_in?
+    redirect_to images_path, flash: { success: 'Logged out' }
```
Comment (bboe) [inline app/controllers/sessions_controller.rb]: Should this be in a separate commit.

File: test/controllers/sessions_controller_test.rb [hunk 11]
```diff
@@ -56,49 +54,52 @@ class SessionsControllerTest < ActionDispatch::IntegrationTest
...
     assert_nil session[:user_id]
     delete session_path
     assert_redirected_to images_path
-    assert_equal 'Logged out', flash[:success]
+    assert_equal 'Already logged out', flash[:warning]
   end
 
+  private
+
+  def create_user_and_login(params)
```
Comment (bboe) [inline test/controllers/sessions_controller_test.rb]: Awesome! Can you fix this up into the commit that it should be introduced with?

File: app/helpers/sessions_helper.rb (hunk 4 above)
Comment (ozhe) [inline app/helpers/sessions_helper.rb]: @bboe I added test in this commit for remember_user

File: app/helpers/sessions_helper.rb (hunk 3 above)
Comment (ozhe) [inline app/helpers/sessions_helper.rb]: @bboe also added test here.

File: test/controllers/sessions_controller_test.rb (hunk 9 above)
Comment (ozhe) [inline test/controllers/sessions_controller_test.rb]: @bboe [2c7d4aef8598a1b3065059ba95acf908294c8204](url)

Comment (ozhe) [inline test/controllers/sessions_controller_test.rb]: @bboe I also split up into three commits to make it more atomic.

File: test/page_objects/sessions/new_page.rb (hunk 6 above)
Comment (ozhe) [inline test/page_objects/sessions/new_page.rb]: @bboe Do we always combine new page object commit with associated flow test?

Comment (bboe) [inline test/page_objects/sessions/new_page.rb]: Yes, they should be one in the same.

File: app/views/sessions/new.html.erb [hunk 12]
```diff
@@ -5,6 +5,7 @@
     <%= simple_form_for @session_form, url: sessions_path do |f| %>
       <%= f.input :email, label: 'Email' %>
       <%= f.input :password, label: 'Password' %>
+      <%= f.input :remember_me, as: :boolean, inline_label: true %>
```
Comment (bboe) [inline app/views/sessions/new.html.erb]: I think it would make sense to include this in the 'implement remember me commit'

File: test/page_objects/sessions/new_page.rb (hunk 6 above)
Comment (bboe) [inline test/page_objects/sessions/new_page.rb]: In this case, since we didn't add them at the time of adding those pages, its fine for them to be in their own commit.

File: test/flows/users_log_in_test.rb [hunk 13]
```diff
@@ -0,0 +1,32 @@
+require 'flow_test_helper'
```
Comment (bboe) [inline test/flows/users_log_in_test.rb]: As per the discussion we're having the page objects likely belong with this commit.
